Wikipedia/Code/Session.swift

```diff
@@ -221,16 +221,24 @@ import Foundation
     }
     
     /**
-     Shared response handling for common status codes. Currently logs the user out and removes local credentials if a 401 is received.
+     Shared response handling for common status codes. Currently logs the user out and removes local credentials if a 401 is received
+     and an attempt to re-login with stored credentials fails.
     */
     private func handleResponse(_ response: URLResponse?) {
         guard let response = response, let httpResponse = response as? HTTPURLResponse else {
             return
         }
         switch httpResponse.statusCode {
         case 401:
-            WMFAuthenticationManager.sharedInstance.logout(initiatedBy: .server) {
-                self.removeAllCookies()
+            WMFAuthenticationManager.sharedInstance.attemptLogin { (loginResult) in
```

Comment:
If this `attemptLogin` call returns a `401` will it get stuck infinitely attempting to login?

Reply:
that's a good point 😂

Reply:
Though I suppose the previous `logout` call had the same potential issue? But I guess it would be uncommon to have `unauthorized` `logout` response...

Reply:
I pushed a WIP to my fork which passes a flag through to session...
https://github.com/wikimedia/wikipedia-ios/commit/f09d0a5c6b53f7f7aac4933114dc19ebc288c62a
...but it feels all sorts of horrible given how many intermediate methods I had to change to relay the value :(

Also, since I don't have time tonight to properly test it, adding `changes welcome`.

Reply:
Sorry I ran out of time before week off!

Reply:
Per the comment above this commit e6f1307 should prevent the `401` recursion 🤞